extra/plugins/dioptra_optic/attacks.py

```diff
@@ -0,0 +1,241 @@
+            :py:data:`numpy.inf`, `1`, or `2`. The default is :py:data:`numpy.inf`.
+        confidence:
+        learning_rate:
+        max_iter:
+        initial_const:
+        save_dataset: Whether to save the dataset to disk.
+
+    Returns:
+        The transformed tf.data.Dataset.
+    """
+    batch_size, num_classes = next(iter(dataset))[1].shape
+    clip_values = (0.0, 255.0) if ops.max(next(iter(dataset))[0]) > 1.0 else (0.0, 1.0)
+
+    method = {
+        "0": CarliniL0Method,
+        "2": CarliniL2Method,
+        "inf": CarliniLInfMethod,
+    }.get(str(norm), None)
+
+    if method is None:
+        raise ValueError("Invalid value for norm, must be one of 0, 2, inf")
+
+    attack = method(
+        classifier=KerasClassifier(model, clip_values=clip_values),
+        confidence=confidence,
+        targeted=target is not None,
+        learning_rate=learning_rate,
+        batch_size=batch_size,
+        max_iter=max_iter,
+        initial_const=initial_const,
+        verbose=False,
+    )
+
+    y_target = (
+        None
+        if target is None
+        else ops.repeat([ops.one_hot(target, num_classes)], batch_size, axis=0)
+    )
+
+    @tf.numpy_function(Tout=(tf.float32, tf.float32))
+    def attack_fn(x, y):
+        return attack.generate(x, y=y_target), y
+
+    return create_transformed_dataset(dataset, attack_fn, save_dataset)
+
+
+@pyplugs.register
+def create_adversarial_patch(
+    model: keras.Model,
+    dataset: tf.data.Dataset,
+    target: int,
+    rotation_max: float = 180.0,
+    scale_min: float = 0.1,
+    scale_max: float = 1.0,
+    learning_rate: float = 5.0,
+    max_iter: int = 500,
+) -> tuple[ArrayLike, ArrayLike]:
+    """Optimizes an Adversarial Patch via expectation over transformation.
+
+    Args:
+        model: The classifier used to generate the patch.
+        dataset: The dataset used to generate the patch.
+        target: The index of target class for the patch.
+        rotation_max: The max angle to rotate the patch by.
+        scale_min: The min scale of the patch as a fraction of image size.
+        scale_max: The max scale of the patch as a fraction of image size.
+        learning_rate: The learning rate used in optimization.
+        max_iter: The maximum iterations in the optimization.
+
+    Returns:
+        The trained patch and its mask.
+    """
+    batch_size, num_classes = next(iter(dataset))[1].shape
+    clip_values = (0.0, 255.0) if ops.max(next(iter(dataset))[0]) > 1.0 else (0.0, 1.0)
+
+    attack = AdversarialPatchNumpy(
+        classifier=KerasClassifier(model, clip_values=clip_values),
+        batch_size=batch_size,
+        rotation_max=rotation_max,
+        scale_min=scale_min,
+        scale_max=scale_max,
+        learning_rate=learning_rate,
+        max_iter=max_iter,
+        verbose=False,
+    )
+
+    x, _ = next(iter(dataset))
+    y_target = ops.repeat([ops.one_hot(target, num_classes)], batch_size, axis=0)
```

Comment:
Isn't this just grabbing the first image return by the tf.data.Dataset object? Is that actually what we would want?